### PR TITLE
Signup: Add type information to the signup controller

### PR DESCRIPTION
**Pull request by p-jackson**
#### Changes proposed in this Pull Request

* Converted `SignupFlowController` to an ES6 class
* Adds types to the `SignupFlowController` class
* Create preliminary types for `Step`, `Flow`, and `Dependencies`. These will move and be expanded on as we add more typing to the signup code.
* Add a few progress store types, since it's used by the controller

Runtime changes made:

##### Removed the `instanceof` check in the constructor

https://github.com/Automattic/wp-calypso/blob/58858cfe98ece9069716374c0638d1509491d483/client/lib/signup/flow-controller.js#L53

This is an old-school pattern meant the "constructor" function could be created with or without the `new` keyword. Not needed now that classes are a real thing, and updated all references to the constructor to make sure `new` is being used.

##### Add null check before calling a step's `apiRequestFunction`

https://github.com/Automattic/wp-calypso/blob/75f093a04006ab9f0d28489eeb490902e4b71941/client/lib/signup/flow-controller.ts#L303

Looks like this fixes a real bug, not just satisfying the type checker. The `apiRequestFunction` property is optional and there wasn't any code checking it existed before calling it.

##### Initialize `this._flow` and `this._flowName` in constructor

These two properties used to be initialised by calling `changeFlowName()`, which is _always_ called in the constructor. If the compiler was clever enough it would be able to figure out that this mean `_flow` and `_flowName` were always initialised, but alas the compiler doesn't trace that far.

One option to address this would have been to leave the runtime behaviour the same and [use the `!` operator to tell the compiler to trust us](https://github.com/microsoft/TypeScript/pull/20166). e.g.

```typescript
export default class SignupFlowController {
	// Notice the "!" here
	_flow!: Flow;
	_flowName!: string;
}
```

But that would mean less type checking as the flow controller continues to evolve.

So instead I've just made it so these properties get intialised twice. Keeps it simple and safe, even if it's a little frustrating when you see it.

### CODE REVIEW

**Comment by p-jackson** [comment]
Keeping e2e tests because they did find a bug in what I'd done, but I think they're false negatives. The failing import test ran correctly locally.

**Comment by andrewserong** [comment]
Nice work @p-jackson! This looks much nicer. I don't feel like I have quite the TypeScript knowledge to give this a proper review just yet, but just wanted to say I like the direction and being clearer about the interfaces.

> Initialize this._flow and this._flowName in constructor

To me this is really good, I like seeing it explicitly set in the constructor there.

The only concern I had otherwise is the potential slight overlap between this change and @spen's work in https://github.com/Automattic/wp-calypso/pull/35139/files#diff-9b8826c512c150c2396692c766af5c35 — it looks like that PR only touches the `SignupFlowController` a bit, so I don't imagine it would be too much trouble to resolve, but just thought I'd flag it 🙂

**Comment by p-jackson** [comment]
Bad rebase, apologies for the unnecessary pings

File: client/state/signup/progress/schema.ts [hunk 1]
```diff
@@ -17,3 +16,13 @@ export const schema = {
...
+
+export type ProgressState = {
+	formData: {
+		url: string;
+	};
+	lastUpdated: number;
+	providedDependencies?: string[];
+	status: 'completed' | 'processing' | 'pending' | 'in-progress' | 'invalid';
+	stepName: string;
+}[];
```
[the remark this answers is not in the document]
Comment (p-jackson) [reply client/state/signup/progress/schema.ts]: Yes I ended up going with the two types approach, one for the array and one for the element type.

File: client/state/signup/progress/schema.ts [hunk 2]
```diff
@@ -17,3 +16,13 @@ export const schema = {
 		},
 	},
 };
+
+export type ProgressState = {
```
[the remark this answers is not in the document]
Comment (p-jackson) [reply client/state/signup/progress/schema.ts]: I like your idea of declaring the type in the same place the shape is defined (i.e. the reducer) better. Long term I think it should be there.

I tried moving it but converting `progress/reducer.js` to TS was non-trivial and didn't want this PR to get out of hand. It's using `createReducer()` which is going to be a tricky beast to type :)

File: client/state/signup/progress/schema.ts [hunk 3]
```diff
@@ -17,3 +16,13 @@ export const schema = {
...
+	formData: {
+		url: string;
+	};
+	lastUpdated: number;
+	providedDependencies?: string[];
+	status: 'completed' | 'processing' | 'pending' | 'in-progress' | 'invalid';
```
[the remark this answers is not in the document]
Comment (p-jackson) [reply client/state/signup/progress/schema.ts]: I changed the schema basically so I could watch the console for warnings and verify that these are the only enum values being used. It's correct as far as I could tell.

**Comment by p-jackson** [comment]
I'm going to base this off #35139 since it's changing the shape of the progress state.

**Comment by p-jackson** [comment]
The changes to the progress state were merged in #35049 so I was able to update the type information to match the new format.

After re-running tests they passed on mobile, and the desktop failure is different from the last failure so they're unrelated to this change.

I think this is ready for another look.

**Pull request merged**